snapcraft/plugins/catkin_tools.py

```diff
@@ -0,0 +1,124 @@
+is the installation of and using catkin_tools to build.
+"""
+
+import os
+import logging
+
+import snapcraft.plugins.catkin
+from snapcraft.plugins.catkin import (
+    Compilers,
+)
+
+import snapcraft
+
+logger = logging.getLogger(__name__)
+
+
+class CatkinToolsPlugin(snapcraft.plugins.catkin.CatkinPlugin):
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+        self.build_packages.extend(['python-catkin-tools'])
+
+    def _prepare_build(self):
+        super()._prepare_build()
+
+        self._clean_catkin()
+
+        self._add_catkin_profile()
+
+        self._configure_catkin_profile()
+
+    def _clean_catkin(self):
+        # Clean to prevent conflicts with externally built packages.
+        catkincmd = ['catkin', 'clean', '-y']
+
+        self._run_in_bash(catkincmd)
+
+    def _add_catkin_profile(self):
+        # Overwrite the default catkin profile to ensure builds
+        # aren't affected by profile changes.
+        catkincmd = ['catkin', 'profile', 'add', '-f', 'default']
+
+        self._run_in_bash(catkincmd)
+
+    def _configure_catkin_profile(self):
+        # Use catkin config to set all configurations before running build.
+        catkincmd = ['catkin', 'config']
+
+        # Use the newly created default profile.
+        catkincmd.extend(['--profile', 'default'])
+
+        # Configure catkin tools to use the snapcraft source, build, and
+        # install directories.
+        catkincmd.extend(['--build-space', self.builddir])
+
+        catkincmd.extend(['--source-space', os.path.join(
+            self.builddir, self.options.source_space)])
+
+        catkincmd.extend(['--install', '--install-space', self.rosdir])
+
+        # Add any cmake-args requested from the plugin options.
+        compilers = Compilers(
+            self._compilers_path, self.PLUGIN_STAGE_SOURCES, self.project)
+
+        catkincmd.append('--cmake-args')
+        build_type = 'Release'
+        if 'debug' in self.options.build_attributes:
+            build_type = 'Debug'
+
+        catkincmd.extend([
+            '-DCMAKE_C_FLAGS="$CFLAGS {}"'.format(compilers.cflags),
+            '-DCMAKE_CXX_FLAGS="$CPPFLAGS {}"'.format(compilers.cxxflags),
+            '-DCMAKE_LD_FLAGS="$LDFLAGS {}"'.format(compilers.ldflags),
+            '-DCMAKE_C_COMPILER={}'.format(compilers.c_compiler_path),
+            '-DCMAKE_CXX_COMPILER={}'.format(compilers.cxx_compiler_path),
+            '-DCMAKE_BUILD_TYPE={}'.format(build_type),
+        ])
+
+        catkincmd.extend(self.options.catkin_cmake_args)
+
+        self._run_in_bash(catkincmd)
+
+    def _build_catkin_packages(self):
+        # Nothing to do if no packages were specified
+        if not self.catkin_packages:
+            return
+
+        # Call catkin build.
+        catkincmd = ['catkin', 'build']
+
+        # Prevent notification that the build is complete.
+        catkincmd.append('--no-notify')
```

Comment:
Does this explode on headless systems? Might be kinda useful, really.

Reply:
Not that I'm aware of.  My thought was that it could be confusing to receive a notification that your build is complete when you still need to go through the stage, prime, and snap steps of the build.

Reply:
Yeah that's a good point.